src/lib.rs

```diff
@@ -0,0 +1,414 @@
+        config_builder = config_builder.credentials_provider(Credentials::new(
+            access_key, secret_key, None, None, "Static",
+        ));
+    };
+
+    let config = config_builder.build();
+    Client::from_conf(config)
+}
+
+#[derive(Debug)]
+// ObjectStore implementation for the Amazon S3 API
+pub struct AmazonS3FileSystem {
+    region: Option<String>,
+    endpoint: Option<String>,
+    retry_max_attempts: Option<u32>,
+    api_call_attempt_timeout_seconds: Option<u64>,
+    access_key: Option<String>,
+    secret_key: Option<String>,
+    bucket: String,
+    client: Client,
+}
+
+impl AmazonS3FileSystem {
+    pub async fn new(
+        region: Option<String>,
+        endpoint: Option<String>,
+        retry_max_attempts: Option<u32>,
+        api_call_attempt_timeout_seconds: Option<u64>,
+        access_key: Option<String>,
+        secret_key: Option<String>,
+        bucket: &str,
+    ) -> Self {
+        Self {
+            region: region.clone(),
+            endpoint: endpoint.clone(),
+            retry_max_attempts,
+            api_call_attempt_timeout_seconds,
+            access_key: access_key.clone(),
+            secret_key: secret_key.clone(),
+            bucket: bucket.to_string(),
+            client: new_client(
+                region,
+                endpoint,
+                retry_max_attempts,
+                api_call_attempt_timeout_seconds,
+                access_key,
+                secret_key,
+            )
+            .await,
+        }
+    }
+}
+
+#[async_trait]
+impl ObjectStore for AmazonS3FileSystem {
+    async fn list_file(&self, prefix: &str) -> Result<FileMetaStream> {
+        let objects = self
+            .client
+            .list_objects_v2()
+            .bucket(&self.bucket)
+            .prefix(prefix)
+            .send()
+            .await
+            .map_err(|err| DataFusionError::Internal(format!("{:?}", err)))?
+            .contents()
+            .unwrap_or_default()
+            .to_vec();
+
+        let result = stream::iter(objects.into_iter().map(|object| {
+            Ok(FileMeta {
+                sized_file: SizedFile {
+                    path: object.key().unwrap_or("").to_string(),
+                    size: object.size() as u64,
+                },
+                last_modified: object
+                    .last_modified()
+                    .map(|last_modified| last_modified.to_chrono_utc()),
+            })
+        }));
+
+        Ok(Box::pin(result))
+    }
+
+    async fn list_dir(&self, _prefix: &str, _delimiter: Option<String>) -> Result<ListEntryStream> {
+        todo!()
+    }
+
+    fn file_reader(&self, file: SizedFile) -> Result<Arc<dyn ObjectReader>> {
+        Ok(Arc::new(AmazonS3FileReader::new(
+            self.region.clone(),
+            self.endpoint.clone(),
+            self.retry_max_attempts,
+            self.api_call_attempt_timeout_seconds,
+            self.access_key.clone(),
+            self.secret_key.clone(),
```

Comment:
Have you considered passing in the already provisioned client here instead of recreating a new client in `sync_chunk_reader`?

Reply:
I found that I was getting a deadlock if I did that as I think the AWS SDK (which is async) is bound to the runtime it is created in. Because we have to wrap the `async` functions with a new runtime to make them execute in the `sync_chunk_reader` they need a new client created each time.

Once `chunk_reader` is implemented (`async`) we can simplify the code significantly.

Reply:
Ha got it, it's unfortunate that they don't provide sync version and only supports tokio. But looks like upstream is working on abstracting out the runtime bit. Could you add a comment in the code to mention the reason for this workaround? I think this is something that needs to be refactored in the future.

Reply:
I have this comment https://github.com/datafusion-contrib/datafusion-objectstore-s3/pull/2/files#diff-b1a35a68f14e696205874893c07fd24fdb88882b47c23cc0e0c80a30c7d53759R242 but can add more if needed

Reply:
I think that's good enough, thank you :+1: